design_docs/2023_07_20_dynamic_shader_parameter_amount.md

```diff
@@ -0,0 +1,55 @@
+# Passing variables to the shader
+We need to be able to pass an arbitrary number of textures and parameters to the custom shaders. For parameters, we need to be able to pass some "header parameters" -- constants used in all videos, as well as some per-video data.
+
+### Possible solutions
+Binding runtime-sized texture arrays works on every backend (except wasm, but it doesn't matter):
+```wgsl
+@group(0) @binding(0) var textures: texture_2d_array<f32>;
+```
+
+Binding const-sized uniform/storage buffers works everywhere too. This would work approximately like this in the shaders (ofc the len can be defined in a different way, e.g. as the length of a runtime-sized texture array):
+```wgsl
+struct MyList {
+	// you declare an array with a constant, large size
+	per_video_data: array<MyStruct, 64>
+	// and then in some way specify how much of this buffer is filled
+	per_video_data_len: u32
+}
+
+@group(0) @binding(0) var<uniform> per_video_data: MyList
```

Comment:
Same bind group and binding as before. I think we should specify in this doc what bind groups and bindings will be used (in which one texture, video parameters, output frame pts, etc. will be passed)

Reply:
what do you think of this:
```wgsl
@group(0) @binding(0) var textures: texture_2d_array<f32>;
@group(1) @binding(0) var<uniform> users_custom_parameters: UsersCustomStruct;
// this can include, for example, the timestamp of this frame 
@group(1) @binding(1) var<uniform> parameters_received_from_the_compositor: OurStruct;
@group(2) @binding(0) var sampler_: sampler;
```

Reply:
LGTM